packages/@lwc/ssr-compiler/src/transmogrify.ts

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (c) 2024, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { traverse, builders as b, type NodePath } from 'estree-toolkit';
+import { produce } from 'immer';
+import type { Node } from 'estree';
+import type { Program as EsProgram } from 'estree';
+
+export type TransmogrificationMode = 'sync' | 'async';
+
+interface TransmogrificationState {
+    mode: TransmogrificationMode;
+}
+
+export type Visitors = Parameters<typeof traverse<Node, TransmogrificationState>>[1];
+
+const EMIT_IDENT = b.identifier('$$emit');
+const TMPL_FN_PATTERN = /tmpl($\d+)?/;
```

Comment:
```suggestion
// Rollup may rename variables to prevent shadowing. When it does, it uses the format `foo$0`, `foo$1`, etc.
const TMPL_FN_PATTERN = /tmpl($\d+)?/;
```